File: app/src/test/java/de/bornholdtlee/defaultproject/base/injection/TestApplication.java
package de.bornholdtlee.defaultproject.base.injection;

import de.bornholdtlee.defaultproject.base.BaseApplication;
import de.bornholdtlee.defaultproject.injection.modules.ApplicationModule;

public class TestApplication extends BaseApplication {

    private TestComponent testComponent;

    @Override
    public TestComponent getAppComponent() {
        return testComponent;
    }

    public DaggerTestComponent.Builder initComponent() {
        return DaggerTestComponent.builder()
                .applicationModule(new ApplicationModule(this));
    }

    public void setTestComponent(TestComponent testComponent) {
        this.testComponent = testComponent;
    }
}

File: app/src/test/java/de/bornholdtlee/defaultproject/base/TestApplication.kt
package de.bornholdtlee.defaultproject.base

import de.bornholdtlee.defaultproject.injection.modules.ApplicationModule
import de.bornholdtlee.defaultproject.injection.modules.UtilsModule
import de.bornholdtlee.defaultproject.utils.SharedPreferencesUtils
import org.mockito.Mockito

class TestApplication : BaseApplication() {

    override val appComponent: TestComponent
        get() = DaggerTestComponent.builder()
                .applicationModule(ApplicationModule(this))
                .utilsModule(provideUtilsModule())
                .build()

    private fun provideUtilsModule(): UtilsModule {
        return object: UtilsModule() {
            override fun provideSharedPreferencesUtils(context: BaseApplication): SharedPreferencesUtils {
                return Mockito.spy(SharedPreferencesUtils::class.java)
            }
        }
    }
}
